spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,415 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2013-2015 The Spyder Development Team
+# Copyright © 2015 Daniel Manson (@d1manson)
+# Licensed under the terms of the MIT License
+# (see spyderlib/__init__.py for details)
+
+from __future__ import print_function
+import os
+import os.path as osp
+
+from spyderlib.qt.QtCore import Signal, Qt, QObject, QSize, QEvent, QPoint
+from spyderlib.qt.QtGui import (QVBoxLayout, QHBoxLayout,
+                                QListWidget, QListWidgetItem,
+                                QDialog, QLineEdit, QKeyEvent,
+                                QStyledItemDelegate, QStyleOptionViewItemV4,
+                                QApplication, QTextDocument, QStyle,
+                                QAbstractTextDocumentLayout, QToolButton,
+                                QToolTip)
+
+# Local imports
+from spyderlib.baseconfig import _
+from spyderlib.guiconfig import new_shortcut
+from spyderlib.py3compat import to_text_string
+from spyderlib.utils.qthelpers import get_std_icon
+
+
+def shorten_paths(path_list):
+    """
+    Takes a list of paths and tries to "intelligently" shorten them all.  The
+    aim is to make it clear to the user where the paths differ, as that is
+    likely what they care about.  Note that this operates on a list of paths
+    not individual paths.
+
+    If the path ends in an actual file this will be removed.
+
+    TODO: at the end, if the path is too long, should do a more dumb kind of
+    shortening, but not completely dumb.
+    """
+
+    # convert the path strings to a list of tokens
+    # and start building the new_path using the drive
+    path_list = path_list[:]  # clone locally
+    new_path_list = []
+    for ii, path in enumerate(path_list):
+        drive, path = osp.splitdrive(osp.dirname(path))
+        new_path_list.append(drive + osp.sep)
+        path_list[ii] = [part for part in path.split(osp.sep) if len(part) > 0]
+
+    def recurse_level(level_idx):
+        # if toks are all empty we need not have reucrsed here
+        if all(len(toks) == 0 for _, toks in level_idx.iteritems()):
+            return
+
+        # firstly, find the longest common prefix for all in the level
+        # s = len of longest common prefix
+        _, sample_toks = next(level_idx.iteritems())
+        if len(sample_toks) == 0:
+            s = 0
+        else:
+            for s, sample_val in enumerate(sample_toks):
+                if not all(len(toks) > s and toks[s] == sample_val
+                           for _, toks in level_idx.iteritems()):
+                    break
+
+        # Shorten longest common prefix
+        if s == 0:
+            short_form = ''
+        else:
+            if s == 1:
+                short_form = sample_toks[0]
+            elif s == 2:
+                short_form = sample_toks[0] + os.sep + sample_toks[1]
```

Comment:
@d1manson have you considered using the Elide text of fontmetrics in Qt? it would take care of you of this shortening, without you having to make it...

http://doc.qt.io/qt-5/qfontmetrics.html#elidedText

And the elide types available